### PR TITLE
[TF/PT] Add FAST detection model

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:  (post 0.8.0 release)

- Add FAST model 

Any feedback is welcome :hugs: 

Todo:

- [x] Check with min pooling in target building -> (same results as with a tuned shrink_ratio)
- [x] Check training
- [x] Check inference

Some notes:
- We skip complex reparameterize (brings only ~5% speed up)
- Instead of only crop on the binary mask from logits we use the current post proc (better controlable with bin_thresh and more save filtering for text rich documents)
- The extensive use of interpolation from the reference implementation is reduced to the required ones

### CODE REVIEW

**Comment by felixdittrich92** [comment]
Screenshot from 2024-01-30 08-01-55
Screenshot from 2024-01-30 08-01-34

used PolyLR

**Comment by felixdittrich92** [comment]
**Inference example (model trained only on FUNSD)**

Screenshot from 2024-01-31 16-04-04

**Comment by nikokks** [comment]
Congrats @felixdittrich92  !!

**Comment by felixdittrich92** [comment]
pooling size tuned boxes are now mostly perfect

File: doctr/models/detection/fast/pytorch.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+    Returns:
+    -------
+        text detection architecture
+    """
+    return _fast(
+        "fast_tiny",
+        pretrained,
+        textnet_tiny,
+        ["3", "4", "5", "6"],
+        ignore_keys=["prob_head.2.weight"],
```
Comment (odulcy-mindee) [inline doctr/models/detection/fast/pytorch.py:335]: Why ignore this specific head ?

Comment (felixT2K) [reply doctr/models/detection/fast/pytorch.py:335]: That's needed if you want to train the model with pretrained for KIE where the number of output classes differs :) Otherwise it results in a shape mismatch from state dict (same for all the other archs)

File: doctr/models/detection/fast/tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
+# Copyright (C) 2021-2024, Mindee.
+
+# This program is licensed under the Apache License 2.0.
+# See LICENSE or go to <https://opensource.org/licenses/Apache-2.0> for full license details.
+
+# Credits: post-processing adapted from https://github.com/xuannianz/DifferentiableBinarization
```
[the remark this answers is not in the document]
Comment (felixT2K) [reply doctr/models/detection/fast/tensorflow.py:6]: Mh the post processing from FAST is not useful in our case (because the model was made for Scene text detection in the wild and it simply extracts the contours from the prob_map this will not work for images with lots of close text) So this should be fine it's the post proc from DB (a bit modified)

File: tests/pytorch/test_models_detection_pt.py [hunk 3]
```diff
@@ -71,7 +74,7 @@ def test_detection_models(arch_name, input_shape, output_size, out_prob, train_m
         },
     ]
     loss = model(input_tensor, target)["loss"]
-    assert isinstance(loss, torch.Tensor) and ((loss - out["loss"]).abs() / loss).item() < 1e-1
+    assert isinstance(loss, torch.Tensor) and ((loss - out["loss"]).abs() / loss).item() < 1
```
[the remark this answers is not in the document]
Comment (felixT2K) [reply tests/pytorch/test_models_detection_pt.py:77]: Mh yeah but this assert really depends on the used / combined loss functions and with adding a model without a checkpoint it will be high 😅 
Maybe changing it back after the model is trained or ignore the failing tests ? xD

File: tests/tensorflow/test_models_detection_tf.py [hunk 4]
```diff
@@ -201,6 +207,12 @@ def test_dilate():
             (512, 512, 1),
             marks=pytest.mark.skipif(system_available_memory < 16, reason="to less memory"),
         ),
+        pytest.param(
+            "fast_base",
+            (512, 512, 3),
+            (512, 512, 1),
+            marks=pytest.mark.skipif(system_available_memory < 16, reason="to less memory"),
```
[the remark this answers is not in the document]
Comment (felixT2K) [reply tests/tensorflow/test_models_detection_tf.py:214]: yep :)

File: tests/pytorch/test_models_detection_pt.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (odulcy-mindee) [reply tests/pytorch/test_models_detection_pt.py:77]: Ok, thanks. We can let it like that

File: doctr/models/detection/fast/pytorch.py (hunk 1 above)
Comment (odulcy-mindee) [reply doctr/models/detection/fast/pytorch.py:335]: Ok thanks

File: doctr/models/detection/fast/tensorflow.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (odulcy-mindee) [reply doctr/models/detection/fast/tensorflow.py:6]: OK !

**Review by odulcy-mindee — APPROVED** [review]
Code looks good, I'm waiting for launching new trainings with it ! :D :rocket:

**Comment by ajkdrag** [comment]
Any estimate on when can we use FAST detection model. DB_resnet50 is very slow for my usecase (i need to run e2e ocr in like 1 sec per image) but it takes longer.

**Comment by felixdittrich92** [comment]
> Any estimate on when can we use FAST detection model. DB_resnet50 is very slow for my usecase (i need to run e2e ocr in like 1 sec per image) but it takes longer.

The plan is to do the release next week and afterwards start with the pytorch `fast_tiny` 

With cuda and half precision (`cuda().half()`) it should also be possible to run `db_resnet50` 1sec/image additional you can set a higher batch size for text recognition `rec_bs=1024` for example (depends on your hardware) which would be another speed up.
Maybe `torch.compile` could also be an option but we haven't tested our models with it yet. :)

CC @odulcy-mindee

**Review by odulcy-mindee — APPROVED** [review]
:rocket: 
brent_rambo
